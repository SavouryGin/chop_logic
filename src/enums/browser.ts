export enum Browser {
  Unknown,
  Opera,
  Chrome,
  Safari,
  Firefox,
}
