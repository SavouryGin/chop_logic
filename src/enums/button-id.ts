export enum ButtonID {
  Cancel,
  Navigation,
  ColorTheme,
  Sounds,
  Settings,
  FullScreen,
  Tools,
  Apply,
  Premise,
  Reiteration,
  Replace,
  Delete,
  ImplicationCreation,
  ImplicationDistribution,
  ContradictionRealization,
  ImplicationElimination,
  Assumption,
  NotIntroduction,
  NotElimination,
  AndIntroduction,
  AndElimination,
  OrIntroduction,
  OrElimination,
  ImpliesIntroduction,
  ImpliesElimination,
  EquivIntroduction,
  EquivElimination,
  Shortcut,
  SavePDF,
  ExportXML,
  ImportXML,
  CopyProof,
  PasteProof,
  CutProof,
  Reset,
}
