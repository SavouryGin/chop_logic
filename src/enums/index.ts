export enum Icon {
  Default = 'chop-icon__check',
  Home = 'chop-icon__home',
  Menu = 'chop-icon__menu',
  Enlarge = 'chop-icon__enlarge2',
  Shrink = 'chop-icon__shrink2',
  Settings = 'chop-icon__cog',
  Delete = 'chop-icon__bin2',
  Up = 'chop-icon__circle-up',
  Down = 'chop-icon__circle-down',
  Right = 'chop-icon__circle-right',
  Left = 'chop-icon__circle-left',
  Sound = 'chop-icon__volume-medium',
  NoSound = 'chop-icon__volume-mute2',
  LightMode = 'chop-icon__sun',
  DarkMode = 'chop-icon__contrast',
}

export enum Pages {
  Home,
  NotFound,
  Propositions,
  Predicates,
  TruthTables,
  Syllogisms,
}
