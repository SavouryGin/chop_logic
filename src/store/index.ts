import createSagaMiddleware from 'redux-saga';
import rootReducer from './root-reducer';
import rootSaga from './root-saga';
import { configureStore } from '@reduxjs/toolkit';

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat([sagaMiddleware]),
});

sagaMiddleware.run(rootSaga);
