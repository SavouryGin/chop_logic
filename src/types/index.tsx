export type ClassNames = {
  [key in string]: boolean;
};

export type ClassNameProp = string | ClassNames;

export type ComponentProps = {
  className?: ClassNameProp;
  id?: string;
};

export type SelectEntity = {
  option: string;
  value: string | number;
  [key: string]: string | number | object;
};
