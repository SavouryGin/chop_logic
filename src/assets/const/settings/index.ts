import { SelectEntity } from 'types';

export const languageOptions: SelectEntity[] = [
  { option: { en: 'English', ru: 'Английский' }, value: 'en' },
  { option: { en: 'Russian', ru: 'Русский' }, value: 'ru' },
];

export const externalLinks: { [key in string]: string } = {
  telegram: 'https://telegram.me/savoury_gin',
  gitHub: 'https://github.com/SavouryGin',
  mail: 'mailto:Dmitrii.Suroviagin@gmail.com',
};
