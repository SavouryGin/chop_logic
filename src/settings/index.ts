import { SelectEntity } from 'types';

export const languageOptions: SelectEntity[] = [
  { option: { en: 'English', ru: 'Английский' }, value: 'en' },
  { option: { en: 'Russian', ru: 'Русский' }, value: 'ru' },
];

export const languageStringOptions = ['en', 'ru'];

export const externalLinks: { [key in string]: string } = {
  telegram: 'https://telegram.me/savoury_gin',
  gitHub: 'https://github.com/SavouryGin',
  linkedIn: 'https://www.linkedin.com/in/dmitrii-suroviagin/',
  mail: 'mailto:Dmitrii.Suroviagin@gmail.com',
  facebook: 'https://www.facebook.com/dmitrii.suroviagin',
};
