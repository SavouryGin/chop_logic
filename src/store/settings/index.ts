import { Language } from 'types';
import { PayloadAction, createSlice } from '@reduxjs/toolkit';
import { SettingsFlag, SettingsInitialState } from './interfaces';

export const settingsInitialState: SettingsInitialState = {
  language: 'en',
  flags: {
    isNavigationOpen: false,
    isSidebarOpen: false,
    isDarkMode: false,
    isFullScreen: false,
    isSettingOpen: false,
    isSoundsEnabled: false,
  },
};

export const settingsSlice = createSlice({
  name: 'settings',
  initialState: settingsInitialState,
  reducers: {
    resetSettings: () => {
      return settingsInitialState;
    },

    toggleFlag: (state, action: PayloadAction<SettingsFlag>) => {
      const currentFlagValue = state.flags[action.payload];
      state.flags[action.payload] = !currentFlagValue;
    },

    setLanguage: (state, action: PayloadAction<Language>) => {
      state.language = action.payload;
    },

    setFullScreenFlag: (state, action: PayloadAction<boolean>) => {
      state.flags.isFullScreen = action.payload;
    },

    setUpFlag: (state, action: PayloadAction<{ flag: SettingsFlag; value: boolean }>) => {
      const { flag, value } = action.payload;
      state.flags[flag] = value;
    },
  },
});

export const settingsActions = { ...settingsSlice.actions };
