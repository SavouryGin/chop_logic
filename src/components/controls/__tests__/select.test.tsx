import React from 'react';
import Select from 'components/controls/select';
import renderWithRedux from 'utils/test-utils/render-with-redux';
import userEvent from '@testing-library/user-event';
import { InputID } from 'enums';
import { SelectEntity } from 'types';
import { combineReducers } from '@reduxjs/toolkit';
import { screen } from '@testing-library/react';
import { settingsInitialState, settingsSlice } from 'store/settings';

describe('Select component:', () => {
  const testSelectOptions: SelectEntity[] = [
    { option: { en: 'One', ru: 'Один' }, value: 1, add: 123 },
    { option: { en: 'Two', ru: 'Два' }, value: 2, asdf: 'asdf' },
    { option: { en: 'Three', ru: 'Три' }, value: 3, asdf: {} },
  ];

  const testProps = {
    name: 'test input',
    label: 'test label',
    inputId: InputID.LanguageSelect,
    options: testSelectOptions,
  };

  const mockedReducer = combineReducers({ settings: settingsSlice.reducer });
  const mockedState = { settings: settingsInitialState };

  it('renders the select element with default props', () => {
    renderWithRedux(<Select {...testProps} />, mockedReducer, mockedState);
    const input = screen.getByRole('combobox');
    expect(input).toBeInTheDocument();
    expect(input).toHaveClass('select__field');
    expect(input).toBeEnabled();
    expect(input).toHaveValue(testSelectOptions[0].value.toString());
  });

  it('the select has a label', () => {
    renderWithRedux(<Select {...testProps} />, mockedReducer, mockedState);
    expect(screen.getByLabelText(testProps.label)).toBeInTheDocument();
  });

  it('the select element contains all the options', () => {
    renderWithRedux(<Select {...testProps} />, mockedReducer, mockedState);
    const options = screen.queryAllByRole('option');
    expect(options).toHaveLength(testSelectOptions.length);
    for (let i = 0; i < testSelectOptions.length; i++) {
      expect(options[i]).toHaveTextContent(testSelectOptions[i].option.en);
    }
  });

  it('can be disabled', () => {
    renderWithRedux(<Select {...testProps} isDisabled />, mockedReducer, mockedState);
    expect(screen.getByRole('combobox')).toBeDisabled();
  });

  it('receives the passed default value', () => {
    renderWithRedux(<Select {...testProps} defaultOption={testSelectOptions[2]} />, mockedReducer, mockedState);
    expect(screen.getByRole('combobox')).toHaveValue(testSelectOptions[2].value.toString());
  });

  it('allows the user to select an option', async () => {
    renderWithRedux(<Select {...testProps} />, mockedReducer, mockedState);
    const input = screen.getByRole('combobox');
    await userEvent.selectOptions(input, testSelectOptions[1].option.en);
    expect(input).toHaveValue(testSelectOptions[1].value.toString());
  });

  it('fires the passed onChange handler', async () => {
    const mockChange = jest.fn();
    renderWithRedux(<Select {...testProps} onChange={mockChange} />, mockedReducer, mockedState);
    const input = screen.getByRole('combobox');
    await userEvent.selectOptions(input, testSelectOptions[1].option.en);
    expect(mockChange).toHaveBeenCalledTimes(1);
  });

  it('gets focus on tab press', async () => {
    renderWithRedux(<Select {...testProps} />, mockedReducer, mockedState);
    const input = screen.getByRole('combobox');
    expect(input).not.toHaveFocus();
    await userEvent.tab();
    expect(input).toHaveFocus();
  });

  it('fires the onBlur handler', async () => {
    const mockBlur = jest.fn();
    renderWithRedux(<Select {...testProps} onBlur={mockBlur} />, mockedReducer, mockedState);
    await userEvent.tab();
    await userEvent.tab();
    expect(mockBlur).toHaveBeenCalledTimes(1);
  });

  it('fires the onFocus handler', async () => {
    const mockFocus = jest.fn();
    renderWithRedux(<Select {...testProps} onFocus={mockFocus} />, mockedReducer, mockedState);
    await userEvent.tab();
    expect(mockFocus).toHaveBeenCalledTimes(1);
  });

  it('should match the snapshot', () => {
    const { asFragment } = renderWithRedux(<Select {...testProps} />, mockedReducer, mockedState);
    expect(asFragment()).toMatchSnapshot();
  });
});
