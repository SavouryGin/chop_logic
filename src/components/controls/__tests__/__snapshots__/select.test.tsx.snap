// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component: should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="select"
  >
    <select
      class="select__field"
      id="select_id_1"
      name="test input"
    >
      <option
        value="1"
      >
        One
      </option>
      <option
        value="2"
      >
        Two
      </option>
      <option
        value="3"
      >
        Three
      </option>
    </select>
    <label
      class="label select__label"
      data-testid="label-test-id"
      for="select_id_1"
    >
      <span
        class="label__text"
      >
        test label
      </span>
    </label>
  </div>
</DocumentFragment>
`;
